In [2]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
import patsy
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns



For my final project, I’ve selected the data from the Kaggle Speed Dating Experiment. 

https://www.kaggle.com/annavictoria/speed-dating-experiment

This data “was gathered from participants in experimental speed dating events from 2002-2004. During the events, the attendees would have a four minute "first date" with every other participant of the opposite sex. At the end of their four minutes, participants were asked if they would like to see their date again. They were also asked to rate their date on six attributes: Attractiveness, Sincerity, Intelligence, Fun, Ambition, and Shared Interests.”

“The dataset also includes questionnaire data gathered from participants at different points in the process. These fields include: demographics, dating habits, self-perception across key attributes, beliefs on what others find valuable in a mate, and lifestyle information. See the Speed Dating Data Key document below for details.”

My hypothesis was to find out whether participants would leave speed dating event with a match. Because of gender differences in partner selection (please see my EDA and scientific literature on attraction), my decision was to build two models: model for males and model for females with dependent variable "match.” First model was predicting a match for females while second model was predicting a match for males. 


“The dataset also includes questionnaire data gathered from participants at different points in the process. These fields include: demographics, dating habits, self-perception across key attributes, beliefs on what others find valuable in a mate, and lifestyle information. See the Speed Dating Data Key document below for details.”

The features were assessed at Time 1 (before and right after participating in the event), Time 2 (the day after participating in the event), and Time 3 (3-4 weeks after participating in the event). I only used data from Time 1 since it was relevant to my model, and I wanted to predict the probability of leaving the event with a match. My project has 4 parts:

Part 1. Data Cleaning

Part 2. EDA

Part 3. Feature selection

Part 4. Model selection



# PART 1. DATA CLEANING 

In [3]:
df=pd.read_csv('/Users/Marina/Documents/GA Data Science/projects/projects-capstone/speed_dating_data.csv')

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8378 entries, 0 to 8377
Data columns (total 98 columns):
iid         8378 non-null int64
id          8377 non-null float64
gender      8378 non-null int64
idg         8378 non-null int64
condtn      8378 non-null int64
wave        8378 non-null int64
round       8378 non-null int64
position    8378 non-null int64
positin1    6532 non-null float64
order       8378 non-null int64
partner     8378 non-null int64
pid         8368 non-null float64
match       8378 non-null int64
int_corr    8220 non-null float64
samerace    8378 non-null int64
age_o       8274 non-null float64
race_o      8305 non-null float64
pf_o_att    8289 non-null float64
pf_o_sin    8289 non-null float64
pf_o_int    8289 non-null float64
pf_o_fun    8280 non-null float64
pf_o_amb    8271 non-null float64
pf_o_sha    8249 non-null float64
dec_o       8378 non-null int64
attr_o      8166 non-null float64
sinc_o      8091 non-null float64
intel_o     8072 non-null float64

### 1). Dropping variables with many missing values:

#### positin1 6532 non-null float64; 
#### undergra 4914 non-null object; 
#### mn_sat   3133 non-null object;
#### tuition 3583 non-null object; 
#### income   4279 non-null object; 
#### expnum  1800 non-null float64; 
#### attr4_1     6489 non-null float64
#### sinc4_1     6489 non-null float64
#### intel4_1    6489 non-null float64
#### fun4_1      6489 non-null float64
#### amb4_1      6489 non-null float64
#### shar4_1     6467 non-null float64
#### attr5_1     4906 non-null float64
#### sinc5_1     4906 non-null float64
#### intel5_1    4906 non-null float64
#### fun5_1      4906 non-null float64
#### amb5_1      4906 non-null float64




In [10]:


df.drop(['positin1','undergra','mn_sat','tuition','income','expnum', 'attr4_1', 'sinc4_1','intel4_1','fun4_1','amb4_1','shar4_1','attr5_1', 'sinc5_1','intel5_1','fun5_1','amb5_1'], axis = 1, inplace=True)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8378 entries, 0 to 8377
Data columns (total 81 columns):
iid         8378 non-null int64
id          8377 non-null float64
gender      8378 non-null int64
idg         8378 non-null int64
condtn      8378 non-null int64
wave        8378 non-null int64
round       8378 non-null int64
position    8378 non-null int64
order       8378 non-null int64
partner     8378 non-null int64
pid         8368 non-null float64
match       8378 non-null int64
int_corr    8220 non-null float64
samerace    8378 non-null int64
age_o       8274 non-null float64
race_o      8305 non-null float64
pf_o_att    8289 non-null float64
pf_o_sin    8289 non-null float64
pf_o_int    8289 non-null float64
pf_o_fun    8280 non-null float64
pf_o_amb    8271 non-null float64
pf_o_sha    8249 non-null float64
dec_o       8378 non-null int64
attr_o      8166 non-null float64
sinc_o      8091 non-null float64
intel_o     8072 non-null float64
fun_o       8018 non-null float64

In [12]:

df1 = df.dropna()


### 2). Saving the file with dropped missning values

In [13]:
df1.to_csv(path_or_buf="/Users/Marina/Documents/GA Data Science/projects/projects-capstone/data_no_missing.csv", index=False)

In [110]:
df=pd.read_csv('/Users/Marina/Documents/GA Data Science/projects/projects-capstone/data_no_missing.csv')

In [111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5731 entries, 0 to 5730
Data columns (total 81 columns):
iid         5731 non-null int64
id          5731 non-null float64
gender      5731 non-null int64
idg         5731 non-null int64
condtn      5731 non-null int64
wave        5731 non-null int64
round       5731 non-null int64
position    5731 non-null int64
order       5731 non-null int64
partner     5731 non-null int64
pid         5731 non-null float64
match       5731 non-null int64
int_corr    5731 non-null float64
samerace    5731 non-null int64
age_o       5731 non-null float64
race_o      5731 non-null float64
pf_o_att    5731 non-null float64
pf_o_sin    5731 non-null float64
pf_o_int    5731 non-null float64
pf_o_fun    5731 non-null float64
pf_o_amb    5731 non-null float64
pf_o_sha    5731 non-null float64
dec_o       5731 non-null int64
attr_o      5731 non-null float64
sinc_o      5731 non-null float64
intel_o     5731 non-null float64
fun_o       5731 non-null float64

### 3). Recoding categorocal features. 

In [112]:
df['field_cd'].value_counts()

8.0     1281
10.0     705
5.0      543
13.0     513
1.0      507
9.0      447
3.0      420
11.0     352
6.0      244
7.0      172
15.0     134
2.0      133
4.0      115
14.0      99
18.0      41
12.0      17
16.0       8
Name: field_cd, dtype: int64

In [113]:
df['field_cd']=df['field_cd'].apply(lambda x: str(x))

In [114]:
## RECODING RACE AND FIELD OF STUDY

def field_of_study_category(words):
        if str(words) =='1.0':
            return 'Law'
        elif str(words) =='2.0':
            return 'Math'
        elif str(words) =='3.0':
            return 'Social Science'
        elif str(words) =='4.0':
            return 'Medical Science'
        elif str(words) =='5.0':
            return 'Engineering'
        elif str(words) =='6.0':
            return 'English'
        elif str(words) =='7.0':
            return 'History/Religion/Philosophy'
        elif str(words) =='8.0':
            return 'Business/Econ/Finance'
        elif str(words) =='9.0':
            return 'Education'
        elif str(words) =='10.0':
            return 'Biology/Chemistry/Physics'
        elif str(words) =='11.0':
            return 'Social Work'
        elif str(words) =='12.0':
            return 'Undergrad/undecided'
        elif str(words) =='13.0':
            return 'Political Science'
        elif str(words) =='14.0':
            return 'Film'
        elif str(words) =='15.0':
            return 'Fine Arts/Arts Admin'
        elif str(words) =='16.0':
            return 'Languages'
        elif str(words) =='17.0':
            return 'Architecture'
        else:
            return 'Other'

In [115]:

df['field_cd']=df['field_cd'].map(field_of_study_category);

df['field_cd'].value_counts()

Business/Econ/Finance          1281
Biologicy/Chemistry/Physics     705
Engineering                     543
Political Science               513
Law                             507
Education                       447
Social Science                  420
Social Work                     352
English                         244
History/Religion/Philosophy     172
Fine Arts/Arts Admin            134
Math                            133
Medical Science                 115
Film                             99
Other                            41
Undergrad/undecided              17
Languages                         8
Name: field_cd, dtype: int64

In [117]:
df['career_c'].value_counts()

2.0     1621
7.0     1501
6.0      524
1.0      523
9.0      402
4.0      279
10.0     246
3.0      161
11.0     148
5.0      131
15.0      59
13.0      49
16.0      36
12.0      26
8.0       17
14.0       8
Name: career_c, dtype: int64

In [118]:
df['career_c']=df['career_c'].apply(lambda x: str(x))

In [119]:
def career_category(words):
        if str(words) =='1.0':
            return 'Lawyer'
        elif str(words) =='2.0':
            return 'Academic/Research'
        elif str(words) =='3.0':
            return 'Psychologist'
        elif str(words) =='4.0':
            return 'Doctor/Medicine'
        elif str(words) =='5.0':
            return 'Engineer'
        elif str(words) =='6.0':
            return 'Entertainment'
        elif str(words) =='7.0':
            return 'Banking/Business'
        elif str(words) =='8.0':
            return 'Real State'
        elif str(words) =='9.0':
            return 'Internatonal Affairs'
        elif str(words) =='10.0':
            return 'Undecided'
        elif str(words) =='11.0':
            return 'Social Work'
        elif str(words) =='12.0':
            return 'Spech Pathology'
        elif str(words) =='13.0':
            return 'Politics'
        elif str(words) =='14.0':
            return 'Athletics/Sports'
        elif str(words) =='15.0':
            return 'Other'
        elif str(words) =='16.0':
            return 'Journalism'
        else:
            return 'Architecture'

In [120]:

df['career_c']=df['career_c'].map(career_category);

df['career_c'].value_counts()

Academic/Research       1621
Banking/Business        1501
Entertainment            524
Lawyer                   523
Internatonal Affairs     402
Doctor/Medicine          279
Undecided                246
Psychologist             161
Social Work              148
Engineer                 131
Other                     59
Politics                  49
Journalism                36
Spech Pathology           26
Real State                17
Athletics/Sports           8
Name: career_c, dtype: int64

In [121]:
df['race']=df['race'].apply(lambda x: str(x))

In [122]:
def race_category(words):
        if '1' in words:
            return 'African American'
        elif '2' in words:
            return 'European American'
        elif '3' in words:
            return 'Hispanic American'
        elif '4' in words:
            return 'Asian American'
        elif '5' in words:
            return 'Native American'
        else:
            return 'Other'

In [123]:

df['race']=df['race'].map(race_category);

df['race'].value_counts()

European American    3312
Asian American       1381
Hispanic American     399
Other                 345
African American      294
Name: race, dtype: int64

In [124]:
df['race_o']=df['race_o'].apply(lambda x: str(x))

In [125]:
def race_o_category(words):
        if '1' in words:
            return 'African American'
        elif '2' in words:
            return 'European American'
        elif '3' in words:
            return 'Hispanic American'
        elif '4' in words:
            return 'Asian American'
        elif '5' in words:
            return 'Native American'
        else:
            return 'Other'

In [126]:

df['race_o']=df['race_o'].map(race_o_category);

df['race_o'].value_counts()

European American    3235
Asian American       1397
Hispanic American     425
Other                 390
African American      284
Name: race_o, dtype: int64

In [127]:
df['gender']=df['gender'].apply(lambda x: str(x))

In [128]:
def gender_category(words):
        if '0' in words:
            return 'Female'
        else:
            return 'Male'

In [129]:
df['gender']=df['gender'].map(gender_category);

df['gender'].value_counts()

Female    2999
Male      2732
Name: gender, dtype: int64

In [130]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5731 entries, 0 to 5730
Data columns (total 81 columns):
iid         5731 non-null int64
id          5731 non-null float64
gender      5731 non-null object
idg         5731 non-null int64
condtn      5731 non-null int64
wave        5731 non-null int64
round       5731 non-null int64
position    5731 non-null int64
order       5731 non-null int64
partner     5731 non-null int64
pid         5731 non-null float64
match       5731 non-null int64
int_corr    5731 non-null float64
samerace    5731 non-null int64
age_o       5731 non-null float64
race_o      5731 non-null object
pf_o_att    5731 non-null float64
pf_o_sin    5731 non-null float64
pf_o_int    5731 non-null float64
pf_o_fun    5731 non-null float64
pf_o_amb    5731 non-null float64
pf_o_sha    5731 non-null float64
dec_o       5731 non-null int64
attr_o      5731 non-null float64
sinc_o      5731 non-null float64
intel_o     5731 non-null float64
fun_o       5731 non-null float64

### 4). Saving final cleaned data to a separate file

In [131]:
df.to_csv(path_or_buf="/Users/Marina/Documents/GA Data Science/projects/projects-capstone/clean_data.csv", index=False)

In [18]:
df.describe()

iid           id          idg       condtn         wave  \
count  5731.000000  5731.000000  5731.000000  5731.000000  5731.000000   
mean    277.710173     8.988658    17.362066     1.848020    11.147269   
std     161.166120     5.398828    10.742096     0.359033     6.114094   
min       4.000000     1.000000     1.000000     1.000000     1.000000   
25%     143.000000     5.000000     9.000000     2.000000     7.000000   
50%     276.000000     8.000000    16.000000     2.000000    11.000000   
75%     410.000000    13.000000    26.000000     2.000000    15.000000   
max     552.000000    22.000000    44.000000     2.000000    21.000000   

             round     position        order      partner          pid  \
count  5731.000000  5731.000000  5731.000000  5731.000000  5731.000000   
mean     16.930204     9.195777     8.831269     9.101902   278.742279   
std       4.238980     5.500188     5.411349     5.514545   161.771526   
min       5.000000     1.000000     1.000000     1.000000     1.000000   
25%      15.000000     4.000000     4.000000     4.000000   144.000000   
50%      18.000000     9.000000     8.000000     8.000000   272.000000   
75%      20.000000    13.000000    13.000000    14.000000   413.000000   
max      22.000000    22.000000    22.000000    22.000000   552.000000   

          ...          intel2_1       fun2_1       amb2_1      shar2_1  \
count     ...       5731.000000  5731.000000  5731.000000  5731.000000   
mean      ...         14.261567    18.366563    11.612146    12.002575   
std       ...          6.041440     6.519451     6.860896     6.056225   
min       ...          0.000000     0.000000     0.000000     0.000000   
25%       ...         10.000000    15.000000     5.000000    10.000000   
50%       ...         15.000000    20.000000    10.000000    10.000000   
75%       ...         19.570000    20.000000    15.000000    15.910000   
max       ...         40.000000    50.000000    50.000000    30.000000   

           attr3_1      sinc3_1       fun3_1     intel3_1       amb3_1  \
count  5731.000000  5731.000000  5731.000000  5731.000000  5731.000000   
mean      7.127377     8.309719     7.735299     8.447042     7.611586   
std       1.378153     1.382340     1.577738     1.060366     1.756831   
min       2.000000     2.000000     2.000000     3.000000     2.000000   
25%       7.000000     8.000000     7.000000     8.000000     7.000000   
50%       7.000000     8.000000     8.000000     8.000000     8.000000   
75%       8.000000     9.000000     9.000000     9.000000     9.000000   
max      10.000000    10.000000    10.000000    10.000000    10.000000   

               dec  
count  5731.000000  
mean      0.428023  
std       0.494835  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000  

[8 rows x 73 columns]